src/Bridge.php

```diff
@@ -1,117 +1,109 @@
+        private ?Server $server = null
+    ) {
+        if (str_ends_with($this->rootDir, '/')) {
+            $this->rootDir = substr($this->rootDir, 0, -1);
+        }
+        if (!file_exists($this->rootDir)) {
+            throw new CakeRoadrunnerException(
+                sprintf(
+                    CakeRoadrunnerException::ROOT_DIR_NOT_FOUND,
+                    $this->rootDir
+                )
+            );
+        }
 
-    /**
-     * @var string root path
-     */
-    protected $root;
+        $configDir = "$this->rootDir/config";
 
-    public function __construct($root = null)
-    {
-        $this->root = $root;
-        if ($root === null) {
-            $this->root = dirname(__DIR__, 4);
+        if ($this->application == null && class_exists('\App\Application')) {
+            $this->application = new \App\Application($configDir);
+        } else {
+            throw new CakeRoadrunnerException(CakeRoadrunnerException::APP_INSTANCE_NOT_CREATED);
         }
-    }
 
-    /**
-     * Bootstrap an application
-     *
-     * @param string|null $appBootstrap The environment your application will use to bootstrap (if any)
-     * @param string $appenv
-     * @param bool $debug If debug is enabled
-     */
-    public function bootstrap($appBootstrap, $appenv, $debug)
-    {
-        require $this->root . '/config/requirements.php';
-        require $this->root . '/vendor/autoload.php';
-        $this->application = new \App\Application($this->root . '/config');
-        $this->application->bootstrap();
+        $this->server = $server ?? new Server($this->application);
 
+        require $configDir . '/requirements.php';
+        $this->application->bootstrap();
+        /*
         if ($this->application instanceof \Cake\Core\PluginApplicationInterface) {
             $this->application->pluginBootstrap();
         }
-        $this->server = new \Cake\Http\Server($this->application);
+        */
     }
 
     /**
      * Handle the request and return a response.
      *
-     * @param ServerRequestInterface $request
-     *
-     * @return ResponseInterface
+     * @param \Psr\Http\Message\ServerRequestInterface $request PSR Server Request Interface
+     * @return \Psr\Http\Message\ResponseInterface
      */
-    public function handle(\Psr\Http\Message\ServerRequestInterface $request)
+    public function handle(ServerRequestInterface $request): ResponseInterface
     {
-        $response = new \Cake\Http\Response();
-        $request = $this->convertRequest($request);
-        $middleware = $this->application->middleware(new \Cake\Http\MiddlewareQueue());
-        if ($this->application instanceof \Cake\Core\PluginApplicationInterface) {
-            $middleware = $this->application->pluginMiddleware($middleware);
-        }
+        $middleware = $this->application->middleware(new MiddlewareQueue());
+        $middleware = $this->application->pluginMiddleware($middleware);
```

Comment:
Not sure what purpose this servers. Also it phpstan complains:

```
> phpstan analyse src/
Note: Using configuration file /var/www/personal/cakephp-roadrunner/phpstan.neon.
 4/4 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ------ ------------------------------------------------------------------------------------- 
  Line   Bridge.php                                                                           
 ------ ------------------------------------------------------------------------------------- 
  87     Call to an undefined method Cake\Core\HttpApplicationInterface::pluginMiddleware().  
 ------ ------------------------------------------------------------------------------------- 
```

Reply:
Sidenote: I wonder if a plugin should be added to test_app to verify plugin middleware is ran...

Reply:
The phpstan error is because `pluginMiddleware()` method is part of `PluginApplicationInterface` where as as `Bridge::$application` is typed as just `HttpApplicationInterface`. So you need to wrap the call in a `instanceof PluginApplicationInterface` check. 

This `handle()` seems to do pretty much the same thing that `Server::run()` does. So either try just calling that or copy it and modify as required.

Reply:
As best I can tell, cake internals and various middleware require `Cake\Http\ServerRequest` in a number of places. For instance things like this fail:

```
Error Object
(
    [message:protected] => Call to undefined method Laminas\Diactoros\ServerRequest::clientIp()
    [string:Error:private] => 
    [code:protected] => 0
    [file:protected] => /var/www/personal/cakephp-swagger-bake-demo/vendor/dereuromark/cakephp-setup/src/Middleware/MaintenanceMiddleware.php
    [line:protected] => 48
```

```
{"title":"CakeDC Roadrunner Error","exception":"TypeError","message":"Cake\\Routing\\Router::setRequest(): Argument #1 ($request) must be of type Cake\\Http\\ServerRequest, Laminas\\Diactoros\\ServerRequest given, called in \/var\/www\/personal\/cakephp-swagger-bake-demo\/vendor\/cakephp\/cakephp\/src\/Routing\/Middleware\/RoutingMiddleware.php on line 135","trace":null}
```

I will just clean up some of the cake that converts to cake server request for now.